### PR TITLE
updated headers

### LINKED ISSUE #792: Standardize header sizes
#### Designs and Mockups

See https://projects.invisionapp.com/d/main#/console/21537330/468374061/preview

> This does require a dev to change what is a h1, h2, h3, etc, but it’s needed for consistency. For example the H1 on blue is not currently an H1 and the SYLLABUS is marked an h1, but it should be an h2… https://ocwnext-rc.odl.mit.edu/courses/21g-102-chinese-ii-regular-spring-2015/pages/syllabus/

#### Acceptance Criteria:

- [ ] This should be applied to the coursev2 theme, in a follow up PR to #784 
- [ ] test on a course with instructor insights

### PR DESCRIPTION
**Pull request by Wassaf-Shahzad**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
- [x] Testing
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes #792 

#### What's this PR do?
- Updates heading tag of title to H1
- Updates heading tag of course-title to h2 

 #### How should this be manually tested?

- Verify heading for course-banner and course-title is h1 and h2

#### Any background context you want to provide?
I have updates the headings for the course banner and title however course content headings such as these 
Screenshot 2022-07-15 at 6 09 27 PM

are set by users and the markdown for it also is for a h1

#### Screenshots (if appropriate)
Screenshot 2022-07-15 at 5 56 33 PM

### CODE REVIEW

**Comment by pdpinch** [comment]
Oh, I thought Abdurrehman was going to work on this. In any case, this change should be part of his course-v2 theme, so we can look at it with everything else. Can you move it over there, maybe after he merges #784 ?

**Comment by pdpinch** [comment]
> I have updated the headings for the course banner and title however course content headings such as these
> are set by users and the markdown for it also is for a h1

Oh, that's going to be challenging. Is there any way to transform that in the theme, with hugo?

**Comment by Wassaf-Shahzad** [comment]
> > I have updated the headings for the course banner and title however course content headings such as these
> > are set by users and the markdown for it also is for a h1
> 
> Oh, that's going to be challenging. Is there any way to transform that in the theme, with hugo?

Yes, we can hugo render hooks to update the level for headings in a specific theme. Render hooks for heading were introduced in v0.71

**Comment by pdpinch** [comment]
Why is this blocked?

**Comment by Wassaf-Shahzad** [comment]
> Why is this blocked?

This was blocked by compact design changes

**Review by gumaerc — CHANGES_REQUESTED** [review]
I think that the scope here is supposed to be limited to the `course-v2` theme and we shouldn't be making any changes to the current `course` theme here.  @pdpinch Can you verify?

The reason the issue asks to check Instructor Insights pages is because way back in https://github.com/mitodl/ocw-course-hugo-theme/pull/20 we added a separate bundle for instructor insights pages so that they could have their own isolated CSS bundle that's only applied on Instructor Insights pages.  The invision designs for these pages back then had different font sizes / colors than other pages.  Take a look at `instructor-insights.scss` for these styles, and verify that the styles here match the typography guide.  Thankfully, it looks like whoever came up with that typography guide also gave font sizes in `rem` assuming a base font size of 16px.

**Comment by gumaerc** [comment]
Also @Wassaf-Shahzad this could use a rebase now that https://github.com/mitodl/ocw-hugo-themes/pull/807 is merged

File: course-v2/layouts/pages/_markup/render-heading.html
Comment (MAbdurrehman12) [inline course-v2/layouts/pages/_markup/render-heading.html:5]: This looks fine but I have a concern/question that will it cause confusion for course authors? Like the course author has set `h2` from studio but it's being rendered as `h3` in Hugo?

Another follow-up question: Is this the correct place to cater to this problem? what if we handle it in studio?
Perhaps if @pdpinch has any thoughts on this?

**Review by MAbdurrehman12 — CHANGES_REQUESTED** [review]
Overall it looks fine, since correct headings are being rendered at correct places.
Just a few pointers mentioned below.

Along with this, an important thing here will be to define/re-define font-sizes for different headings (h1, h2, h3 etc) for this `course-v2` theme. Because currently h3, h4, h5 are just 14px and the values/font-sizes of h1, h2, h3 are also incorrect.

I defined these font-sizes for headings in `course-v2.scss` but that needs to be updated as mentioned [here](https://github.com/mitodl/ocw-hugo-themes/blob/270769d6a3c4ce56dea33cc1b4efdaf552c16e70/course-v2/assets/css/course-v2.scss#L131) (btw, once we completely move to `course-v2` theme, then we can move this change from `course-v2.scss` to some more appropriate place)

So I think we should do that too in this PR.
font-sizes for headings are provided [here](https://projects.invisionapp.com/d/main#/console/21537330/468374061/preview), but the issue which I'm seeing is: `16px` has been considered as the `root font size`, where as, for `course-v2` theme, `root font size` is `14px`.

So as mentioned [here](https://projects.invisionapp.com/d/main#/console/21537330/468374061/preview), for heading 1: `2rem` will not be equal to `32px` for `course-v2` theme. `2rem` will be equal to `28px` because `1rem` = `root font size` i.e: `14px`.

Maybe we can get these font-sizes updated? 
CC: @pdpinch

File: course-v2/assets/css/course-v2.scss
Comment (MAbdurrehman12) [inline course-v2/assets/css/course-v2.scss:154]:
```suggestion
    font-size: 1.4rem !important; // 20px
```

**Review by MAbdurrehman12 — CHANGES_REQUESTED** [review]
Looks good!
However, I think we should:
1. Also define font-sizes for `h4` and `h5` as defined in the [design](https://projects.invisionapp.com/d/main#/console/21537330/468374061/preview).
2. Go upto 3 decimal places for rem values because currently, with 2 decimal places, px values are not equal to what we are trying to set.
for example: 
In case of `h1`: `2.28rem` = `31.92px` (2.28x14) but our target value is `32px`. 
and if we do `2.286rem`, it will be equal to `32px`
Though it won't make any noticeable difference but maybe we can do it just for the sake of accuracy.
What do you think?
Sorry, if I'm being a bit paranoid :P

**Review by MAbdurrehman12 — APPROVED** [review]
LGTM 👍

**Review by gumaerc — CHANGES_REQUESTED** [review]
The changes look properly scoped to `course-v2` now but I think we need to look at one last thing which would be instructor insights pages. They have their own style sheet located at `course-v2/assets/css/instructor-insights.scss`. There is some custom header sizing that we applied in the past for these pages:

```
  h2 {
    margin-bottom: 1.25rem !important;
  }

  h3 {
    font-size: $font-normal !important;
  }
```
According to @pdpinch 's comments on the [issue](https://github.com/mitodl/ocw-hugo-themes/issues/792) we may want to take a look at these pages and make sure that the headers are standardized. An example of a course with instructor insights would be `18.s096-fall-2015`. Here is what the instructor insights page looks like on this branch:

image

We should just make sure that the headers look like they're supposed to in this case before we merge this. The rest of the changes look good and are properly scoped to `course-v2`.

**Comment by MAbdurrehman12** [comment]
@gumaerc I think we should remove this styling
```
  h2 {
    margin-bottom: 1.25rem !important;
  }

  h3 {
    font-size: $font-normal !important;
  }
```
from `course-v2/assets/css/instructor-insights.scss` and we'll be good to go! Right?

ps:  I think this PR will also fix another issue https://github.com/mitodl/ocw-hugo-themes/issues/632 so we can also close #632 once this PR is merged.

**Comment by Wassaf-Shahzad** [comment]
@MAbdurrehman12 @gumaerc I updated instructors-insights.css However I had to remove update it for both course v1 and v2 due to how we have configured webpack.

**Comment by pdpinch** [comment]
Can you give me some before and after URLs I can look at, to see this change? I'd like to have an example of a regular interior page, and an Instructor Insight page.

**Comment by pdpinch** [comment]
Also, in courses-v2 we should be following the color styles specified in https://projects.invisionapp.com/d/main?origin=v7#/console/21537330/468374061/preview?scrollOffset=60 -- no more red

**Comment by Wassaf-Shahzad** [comment]
> https://ocw-hugo-themes-course-pr-794--ocw-next.netlify.app

@pdpinch Updated heading colors

File: course-v2/layouts/pages/_markup/render-heading.html
Comment (pdpinch) [reply course-v2/layouts/pages/_markup/render-heading.html:5]: I think I asked for this, but @MAbdurrehman12 makes a good point. What happens if we don't adjust the heading levels? If possible, give me an example of a course in production.

File: course/assets/css/instructor-insights.scss
Comment (pdpinch) [inline course/assets/css/instructor-insights.scss:10]: Can we limit this change to course-v2 ?

**Comment by pdpinch** [comment]
> https://ocw-hugo-themes-course-pr-794--ocw-next.netlify.app/

> @pdpinch Updated heading colors

Sorry, I'm still seeing the red headings in the build at https://ocw-hugo-themes-course-pr-794--ocw-next.netlify.app/ 

Where can I see this in course-v2?

**Comment by gumaerc** [comment]
@pdpinch We'll need to merge https://github.com/mitodl/ocw-hugo-themes/pull/824 before we'll start seeing `course-v2` builds with Github actions / Netlify.

@Wassaf-Shahzad I'm not sure why we need to also remove the changes from the `course` version of `instructor-insights.scss`? I added these lines back in and took a look at the site rendering with `course`:
image

Running again with `course-v2` I get:
image

Can you elaborate on why you believe we need to remove the CSS from both?

**Comment by MAbdurrehman12** [comment]
> Where can I see this in course-v2?

https://ocw-hugo-themes-course-v2-pr-794--ocw-next.netlify.app/

**Comment by Wassaf-Shahzad** [comment]
> @pdpinch We'll need to merge #824 before we'll start seeing `course-v2` builds with Github actions / Netlify.
> 
> @Wassaf-Shahzad I'm not sure why we need to also remove the CSS from the `course` version of `instructor-insights.scss`? I added these lines back in and took a look at the site rendering with `course`: image
> 
> Running again with `course-v2` I get: image
> 
> Can you elaborate on why you believe we need to remove the CSS from both?

@gumaerc i have not idea why it didn’t  work for me before but its working know so i added back the code.

File: course-v2/layouts/pages/_markup/render-heading.html
Comment (gumaerc) [reply course-v2/layouts/pages/_markup/render-heading.html:5]: @pdpinch Here is the PR preview course in RC with the `course-v2` theme: https://live-qa.ocw.mit.edu/course-v2/courses/15-s21-nuts-and-bolts-of-business-plans-january-iap-2014/

Here is the `course-v2` deploy preview for this PR branch: https://ocw-hugo-themes-course-v2-pr-794--ocw-next.netlify.app/

How font sizes are being set currently:
image

After this PR:
image

The course title header is now wrapped in an h1 tag, but it isn't doing anything to the font size because the link inside of it has a font specified in it via CSS. As far as the content area though, it seems like we're getting h2 and h3 appropriately sized like you'd expect. Why are we demoting h2 to h3 in the above code? If we do end up doing this, there's no need in specifying `{{ add .Level 1}}` because that will always evaluate to 3 in this case, because the if statement is limited to the level being 2.

Comment (pdpinch) [reply course-v2/layouts/pages/_markup/render-heading.html:5]: We've got a couple different requirements, but I can't easily tell if we're meeting them all. 

In order of priority: 

- [ ] every pages should have one and only one H1, for accessibility 
- [ ] additional headers (H2, H3, etc) should follow a logical sequence. In particular, we should avoid skipping any levels. Also for accessibility.
- [ ] Size and color should be correct and consistent, including Instructor Insights pages
- [ ] There should be some logical relationship between the headers users set in Studio (H2, H3, etc) and how they are rendered in the published HTML 
- [ ] We don't want to have to change legacy markdown already in studio. 

Does that help answer your question?

Comment (Wassaf-Shahzad) [reply course-v2/layouts/pages/_markup/render-heading.html:5]: @pdpinch Addressed these checkpoints

**Comment by dseaton** [comment]
If I review this PR based solely on #792, then I think this meets the acceptance criteria. However, there seems to be a lot of other considerations. Could we get @gumaerc to give a final approval in the absence of @pdpinch?

**Review by gumaerc — APPROVED** [review]
:+1: 

@dseaton Everything here has been addressed. I only had one concern that we can address in a follow-up if we need to, relating to this thing that @pdpinch said:

> every page should have one and only one H1, for accessibility

While this is true, the H1 element is always the course title. From an accessibility standpoint, wouldn't we want the H1 to be the page title? Maybe I'm incorrect about this, but if this is a point about accessibility I would think that the H1 would be the first element a screen reader dictates to the user, and having that be the same thing every time seems a bit counter-intuitive. Either way, if that's an issue we can address it later. This is now properly scoped and the correct styles are applied where they should be.

**Comment by dseaton** [comment]
Thanks @gumaerc. Let's talk with @pdpinch next week about the H1 issue. @Wassaf-Shahzad you should be good to go now.

**Pull request merged**
